File: SQL/Expressions/SQLCaseExpression.vb

' ___________________________________________________
'
'  © Hi-Integrity Systems 2010. All rights reserved.
'  www.hisystems.com.au - Toby Wicks
' ___________________________________________________
'

Option Strict On
Option Explicit On

Imports System.Linq
Imports System.Collections.Generic

Namespace SQL

    ''' <summary>
    ''' Allows the use of CASE statements. Specifically:
    ''' 
    ''' Simple CASE expression: 
    ''' CASE input_expression 
    ''' WHEN when_expression THEN result_expression [ ...n ] 
    ''' [ ELSE else_result_expression ] 
    ''' END 
    ''' 
    ''' Searched CASE expression:
    ''' CASE
    ''' WHEN Boolean_expression THEN result_expression [ ...n ] 
    ''' [ ELSE else_result_expression ] 
    ''' END
    ''' </summary>
    Public Class SQLCaseExpression
        Inherits SQLExpression

        Public Class CasesCollection

            Private pobjCases As New List(Of [Case])

            Public Sub Add(ByVal objWhenCondition As SQLExpression, ByVal objResult As SQLExpression)

                pobjCases.Add(New [Case](objWhenCondition, objResult))

            End Sub

            Public Sub Add(ByVal objCase As [Case])

                pobjCases.Add(objCase)

            End Sub

            Friend Function SQL(ByVal eConnectionType As Database.ConnectionType) As String

                If pobjCases.Count = 0 Then
                    Throw New InvalidOperationException("There are no cases for the CASE statement")
                End If

                Return String.Join(" ", pobjCases.Select(Function(objCase) objCase.SQL(eConnectionType)).ToArray)

            End Function

        End Class

        Public Class [Case]

            Private pobjWhenCondition As SQLExpression
            Private pobjResult As SQLExpression

            ''' <summary>
            ''' 
            ''' </summary>
            Public Sub New(ByVal objWhenCondition As SQLExpression, ByVal objResult As SQLExpression)

                If objWhenCondition Is Nothing Then
                    Throw New ArgumentNullException("When condition")
                ElseIf objResult Is Nothing Then
                    Throw New ArgumentNullException("Result")
                End If

                pobjWhenCondition = objWhenCondition
                pobjResult = objResult

            End Sub

            Friend Function SQL(ByVal eConnectionType As Database.ConnectionType) As String

                Return "WHEN " & pobjWhenCondition.SQL(eConnectionType) & " THEN " & pobjResult.SQL(eConnectionType)

            End Function

        End Class

        Private pobjInputExpression As SQLExpression
        Private pobjCases As New CasesCollection
        Private pobjElseResult As SQLExpression

        Public Sub New()

        End Sub

        Public Sub New(ByVal objInputExpression As SQLExpression)

            If objInputExpression Is Nothing Then
                Throw New ArgumentNullException
            End If

            pobjInputExpression = objInputExpression

        End Sub

        Public ReadOnly Property Cases As CasesCollection
            Get

                Return pobjCases

            End Get
        End Property

        Public Property ElseResult As SQLExpression
            Get

                Return pobjElseResult

            End Get

            Set(value As SQLExpression)

                pobjElseResult = value

            End Set
        End Property

        Friend Overrides Function SQL(ByVal eConnectionType As Database.ConnectionType) As String

            Dim strSQL As String = "CASE"

            If pobjInputExpression IsNot Nothing Then
                strSQL &= " " & pobjInputExpression.SQL(eConnectionType)
            End If

            strSQL &= " " & pobjCases.SQL(eConnectionType)

            If pobjElseResult IsNot Nothing Then
                strSQL &= " ELSE " & pobjElseResult.SQL(eConnectionType)
            End If

            strSQL &= " END"

            Return strSQL

        End Function

    End Class

End Namespace

File: SQL/Expressions/SQLCaseExpression.cs
// _________________________________________________________________________
//
//  © Hi-Integrity Systems 2010. All rights reserved.
//  www.hisystems.com.au - Toby Wicks
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//	    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// _________________________________________________________________________
//

using System.Collections;
using System;
using System.Data;
using System.Linq;
using System.Collections.Generic;

namespace DatabaseObjects.SQL
{
	/// <summary>
	/// Allows the use of CASE statements. Specifically:
	///
	/// Simple CASE expression:
	/// CASE input_expression
	/// WHEN when_expression THEN result_expression [ ...n ]
	/// [ ELSE else_result_expression ]
	/// END
	///
	/// Searched CASE expression:
	/// CASE
	/// WHEN Boolean_expression THEN result_expression [ ...n ]
	/// [ ELSE else_result_expression ]
	/// END
	/// </summary>
	public class SQLCaseExpression : SQLExpression
	{
		public class CasesCollection : IEnumerable<Case>
		{
			private List<Case> pobjCases = new List<Case>();
				
			public void Add(SQLExpression objWhenCondition, SQLExpression objResult)
			{
				pobjCases.Add(new Case(objWhenCondition, objResult));
			}
				
			public void Add(Case objCase)
			{
				pobjCases.Add(objCase);
			}

			IEnumerator<Case> IEnumerable<Case>.GetEnumerator()
			{
				return pobjCases.GetEnumerator();
			}

			IEnumerator IEnumerable.GetEnumerator()
			{
				return pobjCases.GetEnumerator();
			}
		}
			
		public class Case
		{
			private SQLExpression pobjWhenCondition;
			private SQLExpression pobjResult;
				
			/// <summary>
			///
			/// </summary>
			public Case(SQLExpression objWhenCondition, SQLExpression objResult)
			{
				if (objWhenCondition == null)
					throw new ArgumentNullException("When condition");
				else if (objResult == null)
					throw new ArgumentNullException("Result");
					
				pobjWhenCondition = objWhenCondition;
				pobjResult = objResult;
			}

			public SQLExpression WhenCondition
			{
				get
				{
					return pobjWhenCondition;
				}
			}

			public SQLExpression Result
			{
				get
				{
					return pobjResult;
				}
			}
		}
			
		private SQLExpression pobjInputExpression;
		private CasesCollection pobjCases = new CasesCollection();
		private SQLExpression pobjElseResult;
			
		public SQLCaseExpression()
		{
		}
			
		public SQLCaseExpression(SQLExpression objInputExpression)
		{
			if (objInputExpression == null)
				throw new ArgumentNullException();
				
			pobjInputExpression = objInputExpression;
		}
			
		public CasesCollection Cases
		{
			get
			{
				return pobjCases;
			}
		}
			
		public SQLExpression ElseResult
		{
			get
			{
				return pobjElseResult;
			}
				
			set
			{
				pobjElseResult = value;
			}
		}

		public SQLExpression InputExpression
		{
			get
			{
				return pobjInputExpression;
			}
		}
			
		internal override string SQL(Serializers.Serializer serializer)
		{
			return serializer.SerializeCaseExpression(this);
		}
	}
}
